src/sage/rings/polynomial/multi_polynomial_sequence.py

```diff
@@ -943,13 +943,24 @@ def connected_components(self):
              Polynomial Sequence with 128 Polynomials in 128 Variables,
              Polynomial Sequence with 128 Polynomials in 128 Variables]
         """
-        g = self.connection_graph()
-        C = sorted(g.connected_components())
-
-        P = [[] for _ in range(len(C))]
+        # Use a union-find data structure to encode relationships between
+        # variables, i.e., that they belong to a same polynomial
+        from sage.sets.disjoint_set import DisjointSet
+        DS = DisjointSet(self.variables())
+        L = []  # to avoid calling twice f.variables()
         for f in self:
-            for i,c in enumerate(C):
-                if len(set(f.variables()).difference(c)) == 0:
+            var = f.variables()
```

Comment:
This is not worse than what it was before, but this is calling `self.variables()` and later `f.variables()` for each `f in self`.

Reply:
Actually, I do one call less to `self.variables()`. You forgot the calls that were done in `connection_graph`.

The proposal of @videlec is certainly doable but much more involved and cannot be done in the same file. I don't have enough knowledge in this part of the code to do it.

Reply:
We can definitely postpone the usage of variable indices rather than variables themselves.

Reply:
I made a more concrete proposal for an interface at  #35523. Comments welcome.